util/zstdframe/options.go

```diff
@@ -0,0 +1,211 @@
+// is used to protect against hostile content.
+// By default, there is no limit.
+// This option is ignored when encoding.
+func MaxDecodedSize(maxSize uint64) Option {
+	return maxDecodedSize(maxSize)
+}
+
+type lowMemory bool
+
+func (lowMemory) isOption() {}
+
+// LowMemory specifies that the encoder and decoder should aim to use
+// lower amounts of memory at the cost of speed.
+// By default, more memory used for better speed.
+func LowMemory(low bool) Option { return lowMemory(low) }
+
+// encoderPool is indexed by every possible combination of options:
+//   - 2 bits for 4 possible encoder levels
+//   - 1 bit for whether to produce a checksum
+//   - 1 bit for for low memory usage
+//
+// Most combinations are unused, and so put no load on the Go runtime.
+// A sync.Pool is registered with runtime only upon first use.
+var encoderPools [1 << (2 + 1 + 1)]sync.Pool // *zstd.Encoder
+
+type encoder struct {
+	pool *sync.Pool
+	*zstd.Encoder
+}
+
+func getEncoder(opts ...Option) encoder {
+	level := zstd.SpeedDefault
+	checksum := true
+	lowMem := false
+	for _, opt := range opts {
+		switch opt := opt.(type) {
+		case encoderLevel:
+			level = zstd.EncoderLevel(opt)
+		case withChecksum:
+			checksum = bool(opt)
+		case lowMemory:
+			lowMem = bool(opt)
+		}
+	}
+
+	var index int
+	switch level {
+	case zstd.SpeedFastest:
+		index |= 0b00
+	case zstd.SpeedDefault:
+		index |= 0b01
+	case zstd.SpeedBetterCompression:
+		index |= 0b10
+	case zstd.SpeedBestCompression:
+		index |= 0b11
+	default:
+		panic(fmt.Sprintf("unknown encoder level: %v", level))
+	}
+	if checksum {
+		index |= 1 << 2
+	}
+	if lowMem {
+		index |= 1 << 3
+	}
+
+	pool := &encoderPools[index]
+	enc, _ := pool.Get().(*zstd.Encoder)
+	if enc == nil {
+		enc = must.Get(zstd.NewWriter(nil,
+			// Set concurrency=1 to ensure synchronous operation.
+			zstd.WithEncoderConcurrency(1),
+			// In stateless compression, the data is already in a single buffer,
+			// so we might as well encode it as a single segment,
+			// which ensures that the Frame_Content_Size is always populated,
+			// informing decoders up-front the expected decompressed size.
+			zstd.WithSingleSegment(true),
+			// Ensure strict compliance with RFC 8878, section 3.1.,
+			// where zstandard "is made up of one or more frames".
+			zstd.WithZeroFrames(true),
+			zstd.WithEncoderLevel(level),
+			zstd.WithEncoderCRC(checksum),
+			zstd.WithLowerEncoderMem(lowMem)))
+	}
+	return encoder{pool, enc}
+}
+
+func putEncoder(e encoder) { e.pool.Put(e.Encoder) }
+
+// decoderPool is indexed by every possible combination of options:
+//   - 6 bits for all 64 power-of-two values of DecoderMaxMemory
+//   - 1 bit for whether to verify a checksum
+//   - 1 bit for for low memory usage
+//
+// Most combinations are unused, and so put no load on the Go runtime.
+// A sync.Pool is registered with runtime only upon first use.
+var decoderPools [1 << (6 + 1 + 1)]sync.Pool // *zstd.Decoder
```

Comment:
how does this compare to using a sync.Map with some struct value type key with the max memory size (not just powers of two even!), and various options? small maps are pretty fast and I imagine there won't be many zstd options in a program.

Reply:
`sync.Map` is slower, and I suspect it's because it's using a `map[any]*entry` under the hood. Yay for the lack of a generic version of `sync.Map` still.

I still decided to use it since it's simpler and it avoids all the overly clever bit manipulation stuff.
We can migrate to the generic `sync.Map` when it eventually sees the light of day.

I also tried `syncs.Map` (the one that uses a Go map with a `sync.RWMutex`) and it performed even worse at highly concurrent workloads.

```
name                         old time/op    new time/op    delta
EncodeParallel/zstdframe-24    51.2µs ± 2%    52.9µs ± 1%   +3.34%  (p=0.008 n=5+5)
DecodeParallel/zstdframe-24    10.7µs ± 1%    12.6µs ± 1%  +17.53%  (p=0.008 n=5+5)
```
This is the performance difference of having all 24 cores compress a 256 byte payload.
For larger payloads, the performance penalty is negligible.

Reply:
I filed https://github.com/golang/go/issues/66446 for some potential future improvements.
There's two levels of inefficiencies. First is the use of `any` as the map key. Suppose we fix that with generics, then we're still going through `runtime.mapaccess` instead of `runtime.mapaccess_fast32` for these key structs. I'd rather not re-introduce the bit-twiddling logic I had earlier and have the Go compiler do the right thing.